### DISCUSSION TITLE
Recommended way of managing Kargo resources?

### DISCUSSION THREAD

**Discussion by oscr** [akuity/kargo#2055] [Q&A] [answered]
In the [Quickstart](https://kargo.akuity.io/quickstart/), resources such as `Warehouse`, `Project` etc are created manually with `kargo apply`. 

However in a production scenario where we want to store the resources as code. What are your recommendations for:
1) Where should these resources be stored? In each microservice repo or in a central "kargo resources" repo? 
2) What would the recommended way of reconciling the changes be? Should there be an ArgoCD instance to reconcile the changes?

**Comment by krancour** [comment, marked as answer]
> However in a production scenario where we want to store the resources as code.

That's the right idea. The whole notion behind Kargo's moving parts being CRDs was precisely so that one _could_ gitops their Kargo projects/pipelines.

> Where should these resources be stored? In each microservice repo or in a central "kargo resources" repo?

My sense is this is really going to come down to an individual organization's requirements and preferences. Just like the question of "Do I store my application's manifests alongside its source code or in a separate repo?", I think a lot of people are going to have different opinions on this and will have various reasons for coming down on one side of it or the other.

> What would the recommended way of reconciling the changes be? Should there be an ArgoCD instance to reconcile the changes?

Yes. We're clearly biased toward Argo CD. I think I would suggest something like using the "app of apps" pattern where each of the lower level Apps manages the resources of a single Kargo Project.

**Reply by oscr** [reply]
Thanks for the answer, and sorry for my somewhat slow reply :pray: 

> My sense is this is really going to come down to an individual organization's requirements and preferences. Just like the question of "Do I store my application's manifests alongside its source code or in a separate repo?", I think a lot of people are going to have different opinions on this and will have various reasons for coming down on one side of it or the other.

The Argo CD [best practice](https://argo-cd.readthedocs.io/en/stable/user-guide/best_practices/#separating-config-vs-source-code-repositories) seems to lean towards having it separate: `Using a separate Git repository to hold your Kubernetes manifests, keeping the config separate from your application source code, is highly recommended for the following reasons`. 

If I understand this correctly, a Kargo `Warehouse` can have one or more git subscriptions. Which means if you have a large number of repos, then Kargo will have to continuously poll all these repos. Because it doesn't support webhooks yet(?). Is this correct? So at this point using a single repo would be a safer bet from a performance perspective.

> Yes. We're clearly biased toward Argo CD. I think I would suggest something like using the "app of apps" pattern where each of the lower level Apps manages the resources of a single Kargo Project.

I think this sounds like a good approach. Maybe this is something which should added to the Kargo docs? Would you be open for a draft pr within this area?
